packages/core/src/provider/hardhat-provider.ts

```diff
@@ -0,0 +1,1268 @@
+/**
+ * This class is a fork of the Nomic Foundation HardhatEthersProvider, credit to them:
+ * https://github.com/NomicFoundation/hardhat/blob/3e32ba0e16bdbf52b8aeb7e696461a221a435dc2/packages/hardhat-ethers/src/internal/hardhat-ethers-provider.ts
+ * We've also forked the associated HardhatEthersSigner class.
+ *
+ * Sphinx needs in-process network for its internal simulation step. This network is created using the `createProvider` function of hardhat which
+ * creates an isolated network suitable for this use case:
+ * https://github.com/NomicFoundation/hardhat/blob/3e32ba0e16bdbf52b8aeb7e696461a221a435dc2/packages/hardhat-core/src/internal/core/providers/construction.ts#L49
+ *
+ * The `createProvider` function returns an `EthereumProvider` object which is not immediately compatible with our execution logic which uses ethers. So to make
+ * the execution process work with this provider, we must convert it a compatible provider. For that purpose, we've created this fork of the `HardhatEthersProvider`.
+ * The `HardhatEthersProvider` is perfect for our use case since it is designed to make the in-process hardhat node compatible with ethers. However, the standard
+ * implementation requires a hardhat project to be used which is problematic for our situation.
+ *
+ * So this fork makes a minor modification to the `HardhatEthersProvider` and `HardhatEthersSigner` classes to allow them to be used outside of a hardhat project.
+ * Where the original implementation relied on reading the hardhat configuration options directly from the `HardhatRuntimeEnvironment` global var, this implementation
+ * accepts a `network` object in the `InProcessEthersProvider` constructor which includes all of the necessary configuration options.
+ *
+ * To review the exact changes we made to this class, you can search `Sphinx` in this file and in the associated `hardhat-signer.ts` file.
+ *
+ * You might wonder why we made a full copy of the both classes just to make these modifications. The reason is that `HardhatEthersSigner` only implements a private
+ * constructor which makes it impossible for us to use subclasses to make the required changes.
+ */
```

Comment:
Please read the whole comment carefully.

I don't recommend reviewing much else in this file, as it's just a copy of the HardhatEthersProvider with minimal modifications. I've left comments indicating my specific changes, though, if you'd like to look through those quickly.